### PR TITLE
Refactor drivers into a separate package out of @daffodil/state

### PR DESCRIPTION
**Pull request by damienwebdev**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[x] Build related changes
[x] CI related changes
[x] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Currently, the API interactions in `@daffodil` are tightly bound into `@daffodil/state`, making it difficult to add code atomically to the driver. Additionally, the API definitions for the driver are sparse and poorly defined.

## What is the new behavior?
This PR separates the API drivers into another package called `@daffodil/driver` and attempts to standardize the API interfaces.

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

To upgrade your own project, you'll need to either install the new driver package alongside the other @daffodil packages, or, if you're working on developing @daffodil, you'll need to run `ng build driver`.

### CODE REVIEW

**Comment by damienwebdev** [comment]
Note that this PR is built on top of the changes done in #35, and will be rebased on develop once that is merged in.

**Comment by damienwebdev** [comment]
https://github.com/angular/angular-cli/issues/10170#issuecomment-380673276

**Comment by damienwebdev** [comment]
@lderrickable this is partially ready for review if you'd like to pull it down and play with it. The only driver implementations thus far are the "testing" and the "in-memory" driver, but the general driver module should provide us with the scaffold for the other drivers fairly easily.

File: libs/driver/in-memory/src/cart/in-mem/cart.testing.service.spec.ts [hunk 1]
```diff
@@ -1,13 +1,10 @@
...
-import { CartTestingModule } from '../../testing/cart-testing.module';
+import { STATUS } from 'angular-in-memory-web-api';
+import { CartFactory } from '@daffodil/core/testing';
+import { DaffInMemoryCartTestingService } from './cart.testing.service';
 
-describe('State | Cart | Testing | CartTestingService', () => {
-  
+describe('Driver | Cart | In Memory | CartTestingService', () => {
   let cartTestingService;
   let cartFactory: CartFactory = new CartFactory();
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/driver/in-memory/src/cart/in-mem/cart.testing.service.spec.ts]: Or maybe this is just something we should do in the future?

File: libs/core/karma.conf.js [hunk 2]
```diff
@@ -25,8 +25,8 @@ module.exports = function (config) {
       colors: true,
       logLevel: config.LOG_INFO,
       autoWatch: true,
-      browsers: ['ChromeHeadless'],
-      singleRun: true
+      browsers: ['Chrome'],
+      singleRun: false
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/core/karma.conf.js]: Post review, I realized you did this everywhere, so I'm guessing this was intentional?

[the remark this answers is not in the document]
Comment (damienwebdev) [reply libs/core/karma.conf.js]: Yeah. I figure, I can set CLI args for the build process. Generally speaking, it's helpful to be able to see the output in the browser while developing.

File: libs/driver/in-memory/src/cart/in-mem/cart.testing.service.spec.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (damienwebdev) [reply libs/driver/in-memory/src/cart/in-mem/cart.testing.service.spec.ts]: @lderrickable I resolved this in 08294d3a065f7fe75085c13ba661fc9e13f4b1c3

File: libs/driver/in-memory/src/in-memory.driver.spec.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import { TestBed } from '@angular/core/testing';
+
+describe('Driver | Shopify | DaffInMemoryDriver', () => {
+  
+});
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply libs/driver/in-memory/src/in-memory.driver.spec.ts]: Removed in 321fe549777b3a69f3b612b2924d3b08c1091a0c

File: libs/state/src/product/effects/product-grid.effects.spec.ts [hunk 4]
```diff
@@ -4,41 +4,37 @@ import { Observable } from 'rxjs/Observable';
 import { of } from 'rxjs/observable/of';
 import { hot, cold } from 'jasmine-marbles';
 
-import { Product, ProductFactory, DaffodilConfigFactory } from '@daffodil/core';
+import { Product } from '@daffodil/core';
+import { ProductFactory } from '@daffodil/core/testing';
+import { DaffProductServiceInterface, DaffDriver, DaffDriverInterface } from '@daffodil/driver';
+import { DaffDriverTestingModule } from '@daffodil/driver/testing';
+
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply libs/state/src/product/effects/product-grid.effects.spec.ts]: @lderrickable  I'm not seeing this. I only see two unused imports. I've cleaned this up in 258075fc43b866ef5b252e52c2d121ec50313780

File: libs/driver/testing/src/cart/services/cart.service.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { Injectable } from '@angular/core';
+
+import { Observable, of } from 'rxjs';
+
+import { Cart } from '@daffodil/core';
+
+import { DaffDriverTestingModule } from '../../testing.module';
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply libs/driver/testing/src/cart/services/cart.service.ts]: I resolved this in 8039347bbbefb9772c1cea4a66043f5829a8a06d

File: libs/driver/shopify/src/product/services/product.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    productService = TestBed.get(DaffShopifyProductService);
+    productFactory = TestBed.get(ProductFactory);
+  });
+
+  it('should be created', () => {
+    expect(productService).toBeTruthy();
+  });
+
+  describe('getAll | getting a list of products', () => {
+    it('should should return an observable array of 20 products by default', () => {
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply libs/driver/shopify/src/product/services/product.service.spec.ts]: Yeah, I'm not quite certain how to handle these tests yet. I'm leaning toward waiting until I get further understanding of graphql and the shopify API.

**Pull request merged**